### PR TITLE
[SPARK-44561][PYTHON] Fix AssertionError when converting UDTF output to a complex type

### PR DESCRIPTION
**Pull request by ueshin**
### What changes were proposed in this pull request?

Fixes AssertionError when converting UDTF output to a complex type by ignore assertions in `_create_converter_from_pandas` to make Arrow raise an error.

### Why are the changes needed?

There is an assertion in `_create_converter_from_pandas`, but it should not be applied for Python UDTF case.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Added/modified the related tests.

### CODE REVIEW

**Comment by ueshin** [comment]
cc @xinrong-meng Could you take a look and see if we want to apply this to Arrow Python UDF too?

**Comment by ueshin** [comment]
cc @allisonwang-db

File: python/pyspark/sql/pandas/types.py [hunk 1]
```diff
@@ -763,6 +764,9 @@ def _create_converter_from_pandas(
     error_on_duplicated_field_names : bool, optional
         Whether raise an exception when there are duplicated field names.
         (default ``True``)
+    ignore_unexpected_complex_type_values : bool, optional
+        Whether ignore the case where unexpected values are given for complex types.
```
Comment (allisonwang-db) [inline python/pyspark/sql/pandas/types.py:768]: Can we add a bit more comment to explain what is an unexpected value for a complex type? I.e., provide some concrete examples.

**Review by allisonwang-db — COMMENTED** [review]
Looks good! Thanks for fixing this!

File: python/pyspark/sql/pandas/types.py [hunk 2]
```diff
@@ -781,28 +785,51 @@ def correct_timestamp(pser: pd.Series) -> pd.Series:
     def _converter(dt: DataType) -> Optional[Callable[[Any], Any]]:
 
         if isinstance(dt, ArrayType):
-            _element_conv = _converter(dt.elementType)
-            if _element_conv is None:
-                return None
+            _element_conv = _converter(dt.elementType) or (lambda x: x)
```
[the remark this answers is not in the document]
Comment (ueshin) [reply python/pyspark/sql/pandas/types.py:788]: `_converter(dt.elementType)` will return `None` when any conversion is needed for the type, `elementType` in this case.
Now we check the given value is `Iterable` or not and always try to convert for array type, too, so just use `lambda x: x` for the case.

File: python/pyspark/sql/tests/test_udtf.py [hunk 3]
```diff
@@ -1789,9 +1786,102 @@ def eval(self):
...
-            # TODO(SPARK-44561): fix AssertionError in convert_map and convert_struct
-            # ("x: map<string,int>", None),
-            # ("x: struct<a:int>", None)
+            ("x: map<string,int>", err),
+            ("x: struct<a:int>", err),
+        ]:
+            with self.subTest(ret_type=ret_type):
+                self._check_result_or_exception(TestUDTF, ret_type, expected)
+
+    def test_map_output_type_casting(self):
```
Comment (allisonwang-db) [inline python/pyspark/sql/tests/test_udtf.py:1795]: Actually, can we also add them in BaseUDTFTestsMixin?

**Review by HyukjinKwon — APPROVED** [review]

**Review by allisonwang-db — APPROVED** [review]

**Comment by ueshin** [comment]
Thanks! merging to master/3.5.

**Pull request closed without merging**